cmd/query/app/server_test.go

```diff
@@ -53,10 +62,314 @@ func TestCreateTLSServerError(t *testing.T) {
+		// 		CertPath:     "invalid/path",
+		// 		KeyPath:      "invalid/path",
+		// 		ClientCAPath: "invalid/path",
+		// 	},
+		// 	clientTLS: tlscfg.Options{
+		// 		Enabled:    true,
+		// 		CertPath:   "invalid/path",
+		// 		KeyPath:    "invalid/path",
+		// 		CAPath:     "invalid/path",
+		// 		ServerName: "example.com",
+		// 	},
+		// 	expectError:      false,
+		// 	expectServerFail: false,
+		// },
+		{
+			name: "Should pass with insecure connection",
+			serverTLS: tlscfg.Options{
+				Enabled: false,
+			},
+			clientTLS: tlscfg.Options{
+				Enabled: false,
+			},
+			expectError:      false,
+			expectServerFail: false,
+		},
+		{
+			name: "should fail with TLS client to untrusted TLS server",
+			serverTLS: tlscfg.Options{
+				Enabled:  true,
+				CertPath: testCertKeyLocation + "/example-server-cert.pem",
+				KeyPath:  testCertKeyLocation + "/example-server-key.pem",
+			},
+			clientTLS: tlscfg.Options{
+				Enabled:    true,
+				ServerName: "example.com",
+			},
+			expectError:      true,
+			expectServerFail: false,
+		},
+		{
+			name: "should fail with TLS client to trusted TLS server with incorrect hostname",
+			serverTLS: tlscfg.Options{
+				Enabled:  true,
+				CertPath: testCertKeyLocation + "/example-server-cert.pem",
+				KeyPath:  testCertKeyLocation + "/example-server-key.pem",
+			},
+			clientTLS: tlscfg.Options{
+				Enabled:    true,
+				CAPath:     testCertKeyLocation + "/example-CA-cert.pem",
+				ServerName: "nonEmpty",
+			},
+			expectError:      true,
+			expectServerFail: false,
+		},
+		{
+			name: "should pass with TLS client to trusted TLS server with correct hostname",
+			serverTLS: tlscfg.Options{
+				Enabled:  true,
+				CertPath: testCertKeyLocation + "/example-server-cert.pem",
+				KeyPath:  testCertKeyLocation + "/example-server-key.pem",
+			},
+			clientTLS: tlscfg.Options{
+				Enabled:    true,
+				CAPath:     testCertKeyLocation + "/example-CA-cert.pem",
+				ServerName: "example.com",
+			},
+			expectError:      false,
+			expectServerFail: false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			serverOptions := &QueryOptions{TLSHTTP: test.serverTLS}
+			httpServer, err := createHTTPServer(&querysvc.QueryService{}, serverOptions, opentracing.NoopTracer{}, zap.NewNop())
+
+			if test.expectServerFail {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				assert.NotNil(t, httpServer)
+			}
+			httpListener, err := net.Listen("tcp", "localhost:"+fmt.Sprintf("%d", ports.QueryHTTP))
+			if test.expectServerFail {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+
+			if serverOptions.TLSHTTP.Enabled {
+
+				tlsCfg, _ := serverOptions.TLSHTTP.Config()
+				tlsCfg.Rand = rand.Reader
+				tlsHTTPListener := tls.NewListener(httpListener, tlsCfg)
+				go func() {
+					// err
+					_ = httpServer.Serve(tlsHTTPListener)
+
+				}()
+
+				time.Sleep(10 * time.Millisecond) // wait for server to start serving
```

Comment:
below code seems identical to code in the else part

Reply:
It actually isn't.  the class of functions for TLS works on `*Conn` rather than `Conn`.   i.e.  the syntax is `(*Conn) Close()`  not `(Conn Close())`
The net package works on `Conn`.  i.e.  the syntax is `(Conn) Close()`  

even though TLS's Conn implements net.Conn,  we cannot overload in this case and use the same functions.  so,  hence If and else are necessary.  The variables are similar for better readability.

Reply:
the only function you use from `conn` is Close, so you could have if-statement creating the `conn` object via tls or net and assign it to something like `io.Closable`, and not repeating the rest of the statements.

Reply:
I used a variable that can store the `X.Close()` method and invoked it outside.